src/java.base/share/classes/java/time/InstantSource.java

```diff
@@ -0,0 +1,243 @@
+ * As such, an {@code InstantSource} can be used instead of {@link System#currentTimeMillis()}.
+ * <p>
+ * The primary purpose of this abstraction is to allow alternate instant sources to be
+ * plugged in as and when required. Applications use an object to obtain the
+ * current time rather than a static method. This can simplify testing.
+ * <p>
+ * Best practice for applications is to pass an {@code InstantSource} into any method
+ * that requires the current instant. A dependency injection framework is one
+ * way to achieve this:
+ * <pre>
+ *  public class MyBean {
+ *    private InstantSource source;  // dependency inject
+ *    ...
+ *    public void process(Instant endInstant) {
+ *      if (source.instant().isAfter(endInstant) {
+ *        ...
+ *      }
+ *    }
+ *  }
+ * </pre>
+ * This approach allows an alternative source, such as {@link #fixed(Instant) fixed}
+ * or {@link #offset(InstantSource, Duration) offset} to be used during testing.
+ * <p>
+ * The {@code system} factory method provides a source based on the best available
+ * system clock. This may use {@link System#currentTimeMillis()}, or a higher
+ * resolution clock if one is available.
+ *
+ * @implSpec
+ * This interface must be implemented with care to ensure other classes operate correctly.
+ * All implementations must be thread-safe.
+ * <p>
+ * The principal methods are defined to allow the throwing of an exception.
+ * In normal use, no exceptions will be thrown, however one possible implementation would be to
+ * obtain the time from a central time server across the network. Obviously, in this case the
+ * lookup could fail, and so the method is permitted to throw an exception.
+ * <p>
+ * The returned instants from {@code InstantSource} work on a time-scale that ignores leap seconds,
+ * as described in {@link Instant}. If the implementation wraps a source that provides leap
+ * second information, then a mechanism should be used to "smooth" the leap second.
+ * The Java Time-Scale mandates the use of UTC-SLS, however implementations may choose
+ * how accurate they are with the time-scale so long as they document how they work.
+ * Implementations are therefore not required to actually perform the UTC-SLS slew or to
+ * otherwise be aware of leap seconds.
+ * <p>
+ * Implementations should implement {@code Serializable} wherever possible and must
+ * document whether or not they do support serialization.
+ *
+ * @implNote
+ * The implementation provided here is based on the same underlying system clock
+ * as {@link System#currentTimeMillis()}, but may have a precision finer than
+ * milliseconds if available.
+ * However, little to no guarantee is provided about the accuracy of the
+ * underlying system clock. Applications requiring a more accurate system clock must
+ * implement this abstract class themselves using a different external system clock,
+ * such as an NTP server.
+ *
+ * @since 17
+ */
+public interface InstantSource {
```

Comment:
Should not we add `@FunctionalInterface`? I can easily imagine this interface being used in tests where we can define the `InstantSource` with lambdas.

Reply:
`@FunctionalInterface` isn't required for use by lambdas. I wasn't initially convinced using it here was the right choice.

Reply:
Right, it's about signaling that it's safe to be used with a lambda/method-reference, so it's not required but a nice to have, very like @Override.